src/utils/read-rows-resumption.ts

```diff
@@ -0,0 +1,258 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {GetRowsOptions, PrefixRange} from '../table';
+import {ChunkTransformer} from '../chunktransformer';
+import * as protos from '../../protos/protos';
+import {TableUtils} from './table';
+import {google} from '../../protos/protos';
+import {CallOptions, GoogleError, RetryOptions} from 'google-gax';
+import {Mutation} from '../mutation';
+import {BoundData, Filter} from '../filter';
+import {RequestType} from 'google-gax/build/src/apitypes';
+import {
+  DEFAULT_BACKOFF_SETTINGS,
+  isRstStreamError,
+  RETRYABLE_STATUS_CODES,
+} from './retry-options';
+import * as is from 'is';
+
+// This interface contains the information that will be used in a request.
+interface TableStrategyInfo {
+  tableName: string;
+  appProfileId?: string;
+}
+
+// Gets the row keys for a readrows request.
+function getRowKeys(
+  rowKeys: string[],
+  lastRowKey: string | number | true | Uint8Array
+) {
+  // Remove rowKeys already read.
+  return rowKeys.filter(rowKey =>
+    TableUtils.greaterThan(rowKey, lastRowKey as string)
+  );
+}
+
+// Modifies ranges in place based on the lastRowKey to prepare
+// a readrows request.
+function spliceRanges(
+  ranges: PrefixRange[],
+  lastRowKey: string | number | true | Uint8Array
+): void {
+  // Readjust and/or remove ranges based on previous valid row reads.
+  // Iterate backward since items may need to be removed.
+  for (let index = ranges.length - 1; index >= 0; index--) {
+    const range = ranges[index];
+    const startValue = is.object(range.start)
+      ? (range.start as BoundData).value
+      : range.start;
+    const endValue = is.object(range.end)
+      ? (range.end as BoundData).value
+      : range.end;
+    const startKeyIsRead =
+      !startValue ||
+      TableUtils.lessThanOrEqualTo(startValue as string, lastRowKey as string);
+    const endKeyIsNotRead =
+      !endValue ||
+      (endValue as Buffer).length === 0 ||
+      TableUtils.lessThan(lastRowKey as string, endValue as string);
```

Comment:
nit: you can probably move this and `endValue` inside the `if (startKeyIsRead)`, since they're not needed otherwise

Reply:
Yes. This is an improvement, because it shortens execution time. I just copied over what was previously in `table.ts` to have as little change as possible, but this is better code.